src/cdk/a11y/high-contrast-mode/high-contrast-mode-detector.ts

```diff
@@ -0,0 +1,89 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {Platform} from '@angular/cdk/platform';
+import {DOCUMENT} from '@angular/common';
+import {Inject, Injectable} from '@angular/core';
+
+
+/** Set of possible high-contrast mode backgrounds. */
+export enum HighContrastMode {
+  NONE,
+  BLACK_ON_WHITE,
+  WHITE_ON_BLACK,
+}
+
+/** CSS class applied to the document body when in black-on-white high-contrast mode. */
+export const BLACK_ON_WHITE_CSS_CLASS = 'cdk-high-contrast-black-on-white';
+
+/** CSS class applied to the document body when in white-on-black high-contrast mode. */
+export const WHITE_ON_BLACK_CSS_CLASS = 'cdk-high-contrast-white-on-black';
+
+/** CSS class applied to the document body when in high-contrast mode. */
+export const HIGH_CONTRAST_MODE_ACTIVE_CSS_CLASS = 'cdk-high-contrast-active';
+
+/**
+ * Service to determine whether the browser is currently in a high-constrast-mode environment.
+ *
+ * Microsoft Windows supports an accessibility feature called "High Contrast Mode". This mode
+ * changes the appearance of all applications, including web applications, to dramatically increase
+ * contrast.
+ */
+@Injectable({providedIn: 'root'})
+export class HighContrastModeDetector {
+  private _document: Document;
+
+  constructor(private _platfrom: Platform, @Inject(DOCUMENT) document: any) {
+    this._document = document;
+  }
+
+  /** Gets the current high-constrast-mode for the page. */
+  getHighContrastMode(): HighContrastMode {
+    if (!this._platfrom.isBrowser) {
+      return HighContrastMode.NONE;
+    }
+
+    // Create a test element with an arbitrary background-color that is neither black nor
+    // white; high-contrast mode will coerce the color to either black or white. Also ensure that
+    // appending the test element to the DOM does not affect layout by absolutely positioning it
+    const testElement = this._document.createElement('div');
```

Comment:
Should we cache this in a similar way to how we cache the viewport size? It's unlikely for this to change dynamically and people might end up calling it frequently, if they don't know that it can be expensive.